zerver/lib/push_notifications.py

```diff
@@ -258,12 +262,23 @@ def send_apple_push_notification(
     payload_data = dict(modernize_apns_payload(payload_data))
     message = {**payload_data.pop("custom", {}), "aps": payload_data}
 
+    for device in devices:
+        if device.ios_app_id is None:
+            # This should be present for all APNs tokens, as an invariant maintained
+            # by the views that add the token to our database.
+            logger.error(
+                "APNs: Missing ios_app_id for user %s device %s", user_identity, device.token
+            )
+
     async def send_all_notifications() -> Iterable[
         Tuple[DeviceToken, Union[aioapns.common.NotificationResult, BaseException]]
     ]:
         requests = [
             aioapns.NotificationRequest(
-                device_token=device.token, message=message, time_to_live=24 * 3600
+                apns_topic=device.ios_app_id,
+                device_token=device.token,
+                message=message,
+                time_to_live=24 * 3600,
```

Comment:
Yeah, ideally we should be taking that device out of the list to send to. (We should still send to the other devices, so that one device somehow getting into a shouldn't-happen state here doesn't break notifications to the user's other devices.)

I hesitated to do that because modifying the list of devices means more code here, none of which should ever run because this should be an impossible condition… but since it raises this question for the reader, it should be addressed.

Reply:
I could also imagine just replacing `invalid.nonsense` with the existing app ID, and just having a comment that it'll typically be overridden but we have this default value since the library requires it?

Reply:
(This will end up on a followup PR — it's in one of the two commits at the tip that will only make sense after dealing with the auth issue.)